File: flutter_app/lib/solver/expr_tree.dart
import 'formula.dart';

/// Canonical expression tree used to detect arithmetically equivalent formulas.
///
/// Strategy:
/// 1. Build a tree from [Formula] steps
/// 2. Normalize: subtraction→add(neg), division→mul(inv)
/// 3. Flatten associative chains (nested sums/products)
/// 4. Cancel matching term/Neg and factor/Inv pairs
/// 5. Remove identity elements (0 in sums, 1 in products)
/// 6. Sort commutative operand lists
/// 7. Serialize to a canonical string for dedup
sealed class Expr {
  /// Build an expression tree from a [Formula] and the original card values.
  static Expr fromFormula(Formula formula, List<int> originalValues) {
    final nodes = originalValues.map((v) => Num(v.toDouble()) as Expr).toList();
    // Steps are stored in reverse order (outermost operation first)
    final reversed = formula.steps.reversed.toList();
    for (final step in reversed) {
      final a = nodes[step.firstIndex];
      final b = nodes[step.secondIndex];
      nodes[step.firstIndex] = BinOp(step.operation.symbol, a, b);
      nodes.removeAt(step.secondIndex);
    }
    return nodes.first;
  }

  /// Return a canonical string representation for deduplication.
  static String canonicalKey(Formula formula, List<int> originalValues) {
    final tree = fromFormula(formula, originalValues);
    final normalized = _normalize(tree);
    return _serialize(normalized);
  }

  // ── Normalization ──

  static Expr _normalize(Expr e) {
    switch (e) {
      case Num():
        return e;
      case Neg():
        final inner = _normalize(e.child);
        // neg(neg(x)) → x
        if (inner is Neg) return inner.child;
        // neg(0) → 0
        if (inner is Num && inner.value == 0) return Num(0);
        // neg(sum(a,b,...)) → sum(neg(a), neg(b), ...)
        if (inner is Sum) {
          return _normalize(Sum(inner.children.map((c) => Neg(c)).toList()));
        }
        return Neg(inner);
      case Inv():
        final inner = _normalize(e.child);
        // inv(inv(x)) → x
        if (inner is Inv) return inner.child;
        // inv(1) → 1
        if (inner is Num && inner.value == 1) return Num(1);
        // inv(prod(a,b,...)) → prod(inv(a), inv(b), ...)
        if (inner is Prod) {
          return _normalizeProd(
              Prod(inner.children.map((c) => Inv(c)).toList()));
        }
        return Inv(inner);
      case BinOp():
        return _normalizeBinOp(e);
      case Sum():
        return _normalizeSum(e);
      case Prod():
        return _normalizeProd(e);
    }
  }

  static Expr _normalizeBinOp(BinOp e) {
    final a = _normalize(e.left);
    final b = _normalize(e.right);
    switch (e.op) {
      case '+':
        return _normalizeSum(Sum([a, b]));
      case '-':
        // a - b → sum(a, neg(b))
        return _normalizeSum(Sum([a, _normalize(Neg(b))]));
      case '*':
        return _normalizeProd(Prod([a, b]));
      case '/':
        // a / b → prod(a, inv(b))
        return _normalizeProd(Prod([a, _normalize(Inv(b))]));
      default:
        return BinOp(e.op, a, b);
    }
  }

  static Expr _normalizeSum(Sum e) {
    // Flatten nested sums and normalize children
    final flat = <Expr>[];
    for (final child in e.children) {
      final nc = _normalize(child);
      if (nc is Sum) {
        flat.addAll(nc.children);
      } else {
        flat.add(nc);
      }
    }

    // Cancel matching term / Neg(term) pairs
    _cancelNegPairs(flat);

    // Remove additive identity (Num(0))
    flat.removeWhere((c) => c is Num && c.value == 0);

    if (flat.isEmpty) return Num(0);
    if (flat.length == 1) return flat.first;

    // Sort for canonical ordering
    flat.sort((a, b) => _serialize(a).compareTo(_serialize(b)));
    return Sum(flat);
  }

  static Expr _normalizeProd(Prod e) {
    // Flatten nested products and normalize children
    final flat = <Expr>[];
    void addFlat(Expr nc) {
      if (nc is Prod) {
        for (final c in nc.children) {
          addFlat(c);
        }
      } else {
        flat.add(nc);
      }
    }
    for (final child in e.children) {
      addFlat(_normalize(child));
    }

    // Cancel matching factor / Inv(factor) pairs
    _cancelInvPairs(flat);

    // Remove multiplicative identity (Num(1))
    flat.removeWhere((c) => c is Num && c.value == 1);

    if (flat.isEmpty) return Num(1);
    if (flat.length == 1) return flat.first;

    // Sort for canonical ordering
    flat.sort((a, b) => _serialize(a).compareTo(_serialize(b)));
    return Prod(flat);
  }

  // ── Pair cancellation ──

  /// Cancel matching x / Neg(x) pairs in-place by canonical serial.
  static void _cancelNegPairs(List<Expr> children) {
    final negIndices = <int>[];
    for (int i = 0; i < children.length; i++) {
      if (children[i] is Neg) negIndices.add(i);
    }
    final toRemove = <int>{};
    for (final ni in negIndices) {
      if (toRemove.contains(ni)) continue;
      final innerKey = _serialize((children[ni] as Neg).child);
      for (int i = 0; i < children.length; i++) {
        if (i == ni || toRemove.contains(i)) continue;
        if (children[i] is! Neg && _serialize(children[i]) == innerKey) {
          toRemove.addAll([ni, i]);
          break;
        }
      }
    }
    final sorted = toRemove.toList()..sort((a, b) => b.compareTo(a));
    for (final i in sorted) {
      children.removeAt(i);
    }
  }

  /// Cancel matching x / Inv(x) pairs in-place by canonical serial.
  static void _cancelInvPairs(List<Expr> children) {
    final invIndices = <int>[];
    for (int i = 0; i < children.length; i++) {
      if (children[i] is Inv) invIndices.add(i);
    }
    final toRemove = <int>{};
    for (final ii in invIndices) {
      if (toRemove.contains(ii)) continue;
      final innerKey = _serialize((children[ii] as Inv).child);
      for (int i = 0; i < children.length; i++) {
        if (i == ii || toRemove.contains(i)) continue;
        if (children[i] is! Inv && _serialize(children[i]) == innerKey) {
          toRemove.addAll([ii, i]);
          break;
        }
      }
    }
    final sorted = toRemove.toList()..sort((a, b) => b.compareTo(a));
    for (final i in sorted) {
      children.removeAt(i);
    }
  }

  // ── Serialization ──

  static String _serialize(Expr e) {
    switch (e) {
      case Num():
        // Use integer representation when possible
        if (e.value == e.value.truncateToDouble()) {
          return e.value.toInt().toString();
        }
        return e.value.toStringAsFixed(6);
      case Neg():
        return 'neg(${_serialize(e.child)})';
      case Inv():
        return 'inv(${_serialize(e.child)})';
      case BinOp():
        return '(${_serialize(e.left)}${e.op}${_serialize(e.right)})';
      case Sum():
        final parts = e.children.map(_serialize).join(',');
        return 'sum($parts)';
      case Prod():
        final parts = e.children.map(_serialize).join(',');
        return 'prod($parts)';
    }
  }
}

class Num extends Expr {
  final double value;
  Num(this.value);
}

class Neg extends Expr {
  final Expr child;
  Neg(this.child);
}

class Inv extends Expr {
  final Expr child;
  Inv(this.child);
}

class BinOp extends Expr {
  final String op;
  final Expr left;
  final Expr right;
  BinOp(this.op, this.left, this.right);
}

class Sum extends Expr {
  final List<Expr> children;
  Sum(this.children);
}

class Prod extends Expr {
  final List<Expr> children;
  Prod(this.children);
}

File: expr_tree.py
"""
Canonical expression tree for algebraic deduplication of 24-game formulas.

Strategy:
1. Build a tree from Formula steps
2. Normalize: subtraction → sum(neg), division → prod(inv)
3. Flatten associative chains (nested sums/products)
4. Cancel matching term/Neg and factor/Inv pairs
5. Remove identity elements (0 in sums, 1 in products)
6. Sort commutative operand lists
7. Serialize to a canonical string for dedup
"""

from Operations.addition import Addition
from Operations.subtraction import Subtraction
from Operations.multiplication import Multiplication
from Operations.division import Division
from Operations.exp import Exp
from Operations.log import Log
from Operations.sqrt import Sqrt


# --- Node types ---

class Num:
    def __init__(self, value):
        if isinstance(value, float) and value.is_integer():
            value = int(value)
        self.value = value


class Neg:
    def __init__(self, child):
        self.child = child


class Inv:
    def __init__(self, child):
        self.child = child


class BinOp:
    def __init__(self, op, left, right):
        self.op = op
        self.left = left
        self.right = right


class Sum:
    def __init__(self, children):
        self.children = children


class Prod:
    def __init__(self, children):
        self.children = children


# --- Operation class → op string mapping ---

_OP_SYMBOL = {
    Addition: '+',
    Subtraction: '-',
    Multiplication: '*',
    Division: '/',
    Exp: '^',
    Log: 'log',
    Sqrt: 'sqrt',
}


# --- Tree building ---

def build_tree(formula, cards):
    """Build an expression tree from a Formula and the original card values."""
    nodes = [Num(v) for v in cards]
    for step in reversed(formula.steps):
        op = _OP_SYMBOL[step["operation"]]
        first = step["first_ind"]
        second = step["second_ind"]
        nodes[first] = BinOp(op, nodes[first], nodes[second])
        nodes.pop(second)
    return nodes[0]


# --- Serialization ---

def serialize(expr):
    """Return a canonical string representation of the expression."""
    if isinstance(expr, Num):
        return str(expr.value)
    if isinstance(expr, Neg):
        return 'neg(' + serialize(expr.child) + ')'
    if isinstance(expr, Inv):
        return 'inv(' + serialize(expr.child) + ')'
    if isinstance(expr, BinOp):
        return '(' + serialize(expr.left) + expr.op + serialize(expr.right) + ')'
    if isinstance(expr, Sum):
        return 'sum(' + ','.join(serialize(c) for c in expr.children) + ')'
    if isinstance(expr, Prod):
        return 'prod(' + ','.join(serialize(c) for c in expr.children) + ')'


# --- Normalization ---

def normalize(expr):
    """Normalize an expression tree for canonical comparison."""
    if isinstance(expr, Num):
        return expr

    if isinstance(expr, Neg):
        inner = normalize(expr.child)
        if isinstance(inner, Neg):
            return inner.child
        if isinstance(inner, Num) and inner.value == 0:
            return Num(0)
        if isinstance(inner, Sum):
            return normalize(Sum([Neg(c) for c in inner.children]))
        return Neg(inner)

    if isinstance(expr, Inv):
        inner = normalize(expr.child)
        if isinstance(inner, Inv):
            return inner.child
        if isinstance(inner, Num) and inner.value == 1:
            return Num(1)
        if isinstance(inner, Prod):
            return normalize(Prod([Inv(c) for c in inner.children]))
        return Inv(inner)

    if isinstance(expr, BinOp):
        return _normalize_binop(expr)

    if isinstance(expr, Sum):
        return _normalize_sum(expr)

    if isinstance(expr, Prod):
        return _normalize_prod(expr)


def _normalize_binop(expr):
    a = normalize(expr.left)
    b = normalize(expr.right)
    op = expr.op

    # --- Advanced cancellation for non-associative ops ---

    # log[x](x) → 1
    if op == 'log' and serialize(a) == serialize(b):
        return Num(1)

    # sqrt[n](x^n) → x
    if op == 'sqrt' and isinstance(a, BinOp) and a.op == '^':
        if serialize(b) == serialize(a.right):
            return a.left

    # log[b](b^x) → x
    if op == 'log' and isinstance(a, BinOp) and a.op == '^':
        if serialize(b) == serialize(a.left):
            return a.right

    # (sqrt[n](x))^n → x
    if op == '^' and isinstance(a, BinOp) and a.op == 'sqrt':
        if serialize(b) == serialize(a.right):
            return a.left

    # b^(log[b](x)) → x
    if op == '^' and isinstance(b, BinOp) and b.op == 'log':
        if serialize(a) == serialize(b.right):
            return b.left

    # --- Convert arithmetic ops to Sum/Prod ---

    if op == '+':
        return _normalize_sum(Sum([a, b]))
    if op == '-':
        return _normalize_sum(Sum([a, normalize(Neg(b))]))
    if op == '*':
        return _normalize_prod(Prod([a, b]))
    if op == '/':
        return _normalize_prod(Prod([a, normalize(Inv(b))]))

    # Non-arithmetic ops stay as BinOp
    return BinOp(op, a, b)


def _normalize_sum(expr):
    # Flatten nested sums and normalize children
    flat = []
    for child in expr.children:
        nc = normalize(child)
        if isinstance(nc, Sum):
            flat.extend(nc.children)
        else:
            flat.append(nc)

    # Cancel matching term / Neg(term) pairs
    flat = _cancel_neg_pairs(flat)

    # Remove additive identity (Num(0))
    flat = [c for c in flat if not (isinstance(c, Num) and c.value == 0)]

    if len(flat) == 0:
        return Num(0)
    if len(flat) == 1:
        return flat[0]

    flat.sort(key=serialize)
    return Sum(flat)


def _normalize_prod(expr):
    # Flatten nested products and normalize children
    flat = []
    for child in expr.children:
        nc = normalize(child)
        if isinstance(nc, Prod):
            flat.extend(nc.children)
        else:
            flat.append(nc)

    # Cancel matching factor / Inv(factor) pairs
    flat = _cancel_inv_pairs(flat)

    # Remove multiplicative identity (Num(1))
    flat = [c for c in flat if not (isinstance(c, Num) and c.value == 1)]

    if len(flat) == 0:
        return Num(1)
    if len(flat) == 1:
        return flat[0]

    flat.sort(key=serialize)
    return Prod(flat)


def _cancel_neg_pairs(children):
    """Cancel matching x / Neg(x) pairs by canonical serial."""
    positives = []
    negatives = []
    for c in children:
        if isinstance(c, Neg):
            negatives.append(c)
        else:
            positives.append(c)

    remaining_pos = list(positives)
    remaining_neg = list(negatives)

    for neg in list(negatives):
        inner_key = serialize(neg.child)
        for i, pos in enumerate(remaining_pos):
            if serialize(pos) == inner_key:
                remaining_pos.pop(i)
                remaining_neg.remove(neg)
                break

    return remaining_pos + remaining_neg


def _cancel_inv_pairs(children):
    """Cancel matching x / Inv(x) pairs by canonical serial."""
    factors = []
    inverses = []
    for c in children:
        if isinstance(c, Inv):
            inverses.append(c)
        else:
            factors.append(c)

    remaining_fac = list(factors)
    remaining_inv = list(inverses)

    for inv in list(inverses):
        inner_key = serialize(inv.child)
        for i, fac in enumerate(remaining_fac):
            if serialize(fac) == inner_key:
                remaining_fac.pop(i)
                remaining_inv.remove(inv)
                break

    return remaining_fac + remaining_inv


# --- Public API ---

def canonical_key(formula, cards):
    """Return a canonical string key for algebraic deduplication."""
    tree = build_tree(formula, cards)
    normalized = normalize(tree)
    return serialize(normalized)
